cmd/boulder-va/gsb.go

```diff
@@ -5,34 +5,109 @@ import (
 	"fmt"
 	"os"
 
+	safebrowsingv4 "github.com/google/safebrowsing"
 	"github.com/letsencrypt/boulder/cmd"
 	"github.com/letsencrypt/boulder/va"
 	safebrowsing "github.com/letsencrypt/go-safe-browsing-api"
 )
 
-// newGoogleSafeBrowsing returns nil if the GoogleSafeBrowsing struct given is
-// nil. If an empty Google API key or an unreadable data directory is in the
-// GoogleSafeBrowsing config struct, this function runs cmd.FailOnError.
-func newGoogleSafeBrowsing(gsb *cmd.GoogleSafeBrowsingConfig) va.SafeBrowsing {
+var (
+	NilConfigErr   = errors.New("Google Safe Browsing config was nil")
+	EmptyAPIKeyErr = errors.New("a Google Safe Browsing config was given but " +
+		"it did not include a Google API key in APIKey")
+	EmptyDataDirErr = errors.New("a Google Safe Browsing config was given but " +
+		"it did not include a DataDir for persistence")
+	MissingDataDirErr = errors.New("a Google Safe Browsing data directory was " +
+		"given but it does not exist")
+	BadDataDirErr = errors.New("a Google Safe Browsing data directory was " +
+		"given but it cannot be opened")
+)
+
+// configCheck returns an error if:
+// * the gsb config struct given is nil
+// * the gsb config struct's APIKey is empty
+// * the gsb config struct's DataDir is empty
+// * the gsb config struct's DataDir doesn't exist or isn't readable
+func configCheck(gsb *cmd.GoogleSafeBrowsingConfig) error {
 	if gsb == nil {
-		return nil
+		return NilConfigErr
 	}
 	if gsb.APIKey == "" {
-		cmd.FailOnError(errors.New(""), "a Google Safe Browsing config was given but it did not include a Google API key in APIKey")
+		return EmptyAPIKeyErr
 	}
 	if gsb.DataDir == "" {
-		cmd.FailOnError(errors.New(""), "a Google Safe Browsing config was given but it did not include data directory to store the hashes file in DataDir")
+		return EmptyDataDirErr
 	}
-
 	f, err := os.Open(gsb.DataDir)
+	// NOTE: Using `defer f.Close()` instead makes errcheck unhappy.
+	defer func() { _ = f.Close() }()
 	if err != nil {
 		if os.IsNotExist(err) {
-			cmd.FailOnError(err, fmt.Sprintf("Google Safe Browsing data directory (%#v) does not exist", gsb.DataDir))
+			return MissingDataDirErr
 		}
-		cmd.FailOnError(err, "unable to open Google Safe Browsing data directory")
+		return BadDataDirErr
+	}
+	return nil
+}
+
+// gsbAdapter adapts the Google safebrowsing's `SafeBrowser` type to the
+// `va.SafeBrowsing` interface Boulder uses.
+type gsbAdapter struct {
+	*safebrowsingv4.SafeBrowser
+}
+
+// IsListed provides the va.SafeBrowsing interface by using the
+// `safebrowsing4v.SafeBrowser` to look up one URL and return the first threat
+// list it is found on, or "" if the URL is safe.
+func (sb gsbAdapter) IsListed(url string) (string, error) {
+	threats, err := sb.LookupURLs([]string{url})
+	if err != nil {
+		return "error", err
```

Comment:
Seemed like defense-in-depth since `""` is the "This URL is Safe" value and would fail-open if someone forgot to check `err`. I agree it's a little ugly. I could be convinced to switch to `return "", err`